[corechecks/kubelet] Add probe metrics

sblumenthal opened the pull request:
### What does this PR do?

This PR captures and reports metrics from the /metrics/probes kubelet endpoint

### Motivation

Part of the process of moving the kubelet check from integrations-core over here.

### Additional Notes

- This PR implements the probes transformer which is defined [here](https://github.com/DataDog/integrations-core/blob/6ae41f6a47e0154e6c54a1d3658d99fd301927f5/kubelet/datadog_checks/kubelet/probes.py)
- It does this by copying some of the logic from the [openmetrics mixin here](https://github.com/DataDog/integrations-core/blob/6ae41f6a47e0154e6c54a1d3658d99fd301927f5/datadog_checks_base/datadog_checks/base/checks/openmetrics/mixins.py#L761)

### Possible Drawbacks / Trade-offs

The goal here was not to fully implement the openmetrics python mixin, as that is a much larger scope problem and is not fully necessary to support the prometheus functionality which the kubelet check needs. As such, there are a bunch of known omissions at this point which will be implemented in the future, including but not limited to:
- label filtering and inclusion based on config value
- metric omission based on labels
- histogram support
- monotonic counter support
- label mapping
- telemetry

In addition, because of the narrow scope for now, the existing kubelet connection is being used, so support for creating a generic http client would also need to be added.

As the prometheus code continues to be brought up to feature parity, we will probably want to move the bulk of it out of the kubelet check and into some other place. But for now, it is being included here so we can continue to iterate while limiting the potential of it being used elsewhere when it is not ready.

### Describe how to test/QA your changes

Similar to https://github.com/DataDog/datadog-agent/pull/17913:

Deploy the agent image using the public helm chart with the following values:
```
datadog:
  confd:
    kubelet_core.yaml: |
      init_config:
      instances:
        - min_collection_interval: 20
```

and validate that `kubernetes_core.<liveness|readiness|startup>_probe...` metrics are being generated, either via the UI or via:
```
agent check -r kubelet_core
```

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

sblumenthal committed:
Add probe metrics to kubelet corecheck

sblumenthal committed:
Implement some of the python prometheus mixin logic

sblumenthal committed:
Update comments, fix failing test, clean up code

sblumenthal committed:
Add support for startup probe metrics

sblumenthal committed:
Fix error surrounding probes_metrics_endpoint config

clamoriniere commented on pkg/collector/corechecks/containers/kubelet/provider/probe/provider.go:
```diff
@@ -0,0 +1,110 @@
+type Provider struct {
+	filter *containers.Filter
+	store  workloadmeta.Store
+	*prometheus.Provider
```
can you explain why we use a pointer here for this composition?

sblumenthal replied:
No reason, a remnant from when I was moving things around in my branch. The functionality is the same without it, so I will remove it

sblumenthal replied:
I have removed it, thanks for pointing it out to me

clamoriniere commented on pkg/collector/corechecks/containers/kubelet/provider/probe/provider.go:
```diff
@@ -0,0 +1,110 @@
+		store:  store,
+	}
+
+	transformers := map[string]func(*model.Sample, sender.Sender){
```
can we define the transformer function signature outside this function? 

is it something that will be passed to the `provider.Prometheus` ?

clamoriniere replied:
```suggestion
    // type TransformerFuncType func(*model.Sample, sender.Sender)
	transformers := map[string]TransformerFuncType{
```

clamoriniere commented on pkg/collector/corechecks/containers/kubelet/provider/prometheus/provider.go:
```diff
@@ -0,0 +1,197 @@
+	AllowNotFound bool
+}
+
+func NewProvider(config *common.KubeletConfig, transformers map[string]func(*model.Sample, sender.Sender), scraperConfig *ScraperConfig) *Provider {
```
Since NewProvider is used in another package, it can be good to also define what is a Tranformer function signature

```suggestion
type TransformerFuncType func(*model.Sample, sender.Sender)
type TranformersType map[string] TransformerFuncType

func NewProvider(config *common.KubeletConfig, transformers map[string]func(*model.Sample, sender.Sender), scraperConfig *ScraperConfig) *Provider {
```

sblumenthal replied:
Good callout, I had thought about doing it but I wasn't sure if it would be overkill or not. Your comment sort of sealed the deal with it! I have added the change

clamoriniere commented on pkg/collector/corechecks/containers/kubelet/provider/probe/provider.go:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	result := metric.Metric["result"]
+	if result == "successful" {
+		metricSuffix += ".success.total"
+	} else if result == "failed" {
+		metricSuffix += ".failure.total"
+	} else if result == "unknown" {
+		metricSuffix += ".unknown.total"
+	} else {
+		log.Debugf("Unsupported probe result %s", result)
+		return
+	}
```
same about switch case

clamoriniere reviewed (commented):
Nice work. 

some nits about golang and error handeling.

sblumenthal commented on pkg/collector/corechecks/containers/kubelet/provider/probe/provider.go:
```diff
@@ -0,0 +1,110 @@
+		return
+	}
+
+	podUid := string(metric.Metric["pod_uid"])
```
If it is not going to be there for one, it would not be there for all, because it is coming from the output of the /metrics/probes endpoint. If it is missing, the end user would not be able to do anything about it, and I worry that the amount of times we could run into it could result in a lot of log spam if we do log a warn.

If the desire is to have logging for such an instance, then I think having it be debug level makes more sense, so that the ability to debug missing metrics still exists. But in either case I am happy to make the change if you think that the logging volume is not that much of a concern

sblumenthal commented on pkg/collector/corechecks/containers/kubelet/provider/probe/provider.go:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	podUid := string(metric.Metric["pod_uid"])
+	containerName := string(metric.Metric["container"])
```
See my response above

sblumenthal committed:
Refactor slightly based on PR feedback

- Removes pointer from probe.Provider composition
- Defines type aliases for the prometheus transformer function signature
- Use switch case instead of if / else block

sblumenthal committed:
Merge remote-tracking branch 'origin' into sblumenthal/cont-3871-kubelet-core-probes-endpoint

sblumenthal committed:
Fix syntax error after changing function signature

sblumenthal committed:
Remove `\r` from prometheus input to fix windows tests

- Apparently the prometheus TextParser breaks when running on windows,
  because the lines end with "\r\n", but the prometheus code is just looking
  for "\n". It is resulting in errors like:

  `text format parsing error in line 2: unknown metric type "counter\r"`

clamoriniere reviewed (commented):
Awesome work looking forward seeing it running live 🚀

sblumenthal committed:
Merge remote-tracking branch 'origin' into sblumenthal/cont-3871-kubelet-core-probes-endpoint

# Conflicts:
#	go.mod
